rust/kernel/allocator.rs

```diff
@@ -2,29 +2,136 @@
+    layout.align() > bindings::ARCH_KMALLOC_MINALIGN as usize
+        && layout.size() > layout.align()
+        && !is_power_of_two(layout.size())
+}
+
+// Header containing a pointer to the start of an allocated block.
+// SAFETY: Size and alignment must be <= `bindings::ARCH_KALLOC_MINALIGN`.
+#[repr(C)]
+struct Header(*mut u8);
+
+static_assert!(core::mem::size_of::<Header>() <= bindings::ARCH_KMALLOC_MINALIGN);
+static_assert!(core::mem::align_of::<Header>() <= bindings::ARCH_KMALLOC_MINALIGN);
+
+impl KernelAllocator {
+    #[cfg(not(test))]
+    #[cfg(not(testlib))]
+    fn allocate_with_header(
+        &self,
+        layout: Layout,
+        flags: bindings::gfp_t,
+    ) -> Result<NonNull<[u8]>, AllocError> {
+        let alloc_size = layout.size() + layout.align();
+
+        // SAFETY: `krealloc` will return a pointer to `alloc_size` memory block or a null pointer
+        let block = unsafe { bindings::krealloc(ptr::null(), alloc_size, flags) as *mut u8 };
+        if block.is_null() {
+            return Err(AllocError);
+        }
+
+        // Create a correctly aligned pointer offset from the start of the allocated block,
+        // and write a header before it.
+        let offset = layout.align() - (block.addr() & (layout.align() - 1));
+
+        // SAFETY: `bindings::ARCH_KALLOC_MINALIGN` <= `offset` <= `layout.align()` and the size
+        // of the allocated block is `layout.align() + layout.size()`. `aligned` will thus be a
+        // correctly aligned pointer inside the allocated block with at least `layout.size()`
+        // bytes after it and at least `bindings::ARCH_KALLOC_MINALIGN` bytes of padding before
+        // it.
+        let aligned = unsafe { block.add(offset) };
+
+        // SAFETY: Because the size and alignment of a header is <=
+        // `bindings::ARCH_KALLOC_MINALIGN` and `aligned` is aligned to at least
+        // `bindings::ARCH_KALLOC_MINALIGN` and has at least `bindings::ARCH_KALLOC_MINALIGN`
+        // bytes of padding before it, it is safe to write a header directly before it.
+        unsafe { ptr::write((aligned as *mut Header).offset(-1), Header(block)) };
+
+        // SAFETY: The returned pointer does not point to the to the start of an allocated block,
+        // but there is a header readable directly before it containing the location of the start
+        // of the block. We checked that the pointer is not null above.
+        return Ok(unsafe {
+            NonNull::new_unchecked(core::slice::from_raw_parts_mut(aligned, layout.size()))
+        });
+    }
+
+    #[cfg(not(test))]
+    #[cfg(not(testlib))]
+    pub(crate) fn allocate_with_flags(
+        &self,
+        mut layout: Layout,
+        flags: bindings::gfp_t,
+    ) -> Result<NonNull<[u8]>, AllocError> {
+        if requires_header(&layout) {
+            return self.allocate_with_header(layout, flags);
+        } else if layout.size() < layout.align() {
+            layout = layout.pad_to_align();
```

Comment:
Isn't this just setting `size` to `align`?

Reply:
Setting the size requires either `Layout::from_size_align` (which can return an error) or `Layout::from_size_align_unchecked` (which is unsafe). Seems like `pad_to_align` will not optimize into a plain size setting though.

Reply:
`Layout` is not used later anymore so it can be destructured.

Reply:
> `Layout` is not used later anymore so it can be destructured.

Not sure I follow. Why would we explicitly destroy the layout here? Size parameter is used in the allocation :shrug: